core/modules/qdisp/Executive.cc

```diff
@@ -367,8 +355,11 @@ bool Executive::_shouldRetry(int refNum) {
     return true;
 }
 
-ExecStatus& Executive::_insertNewStatus(int refNum,
-                                        ResourceUnit const& r) {
+// Returning a reference instead of a shared pointer is an optimization.
+// It requires that the reference does not outlast the pointer in _statuses.
+// At the time of writing, everything in _statuses lasts until this Executive
+// is deleted.
+ExecStatus& Executive::_insertNewStatus(int refNum, ResourceUnit const& r) {
```

Comment:
I don't know if this kind of optimization is worth it, it may break easily and lead to complex bug in maintenance (whereas I've done it sometime). Maybe we should decide a standard way of doing this: for example only optimize shared pointer argument/return object after monitoring has shown it is worth it?